### PR TITLE
Fix crash in amqp_management if response to operation is received before on_send_complete is invoked.

### PR DESCRIPTION
**Pull request by ewertons**

### CODE REVIEW

File: src/amqp_management.c [hunk 1]
```diff
@@ -343,18 +355,24 @@ static void on_message_send_complete(void* context, MESSAGE_SEND_RESULT send_res
     }
     else
     {
+        /* Codes_SRS_AMQP_MANAGEMENT_01_168: [ - `context` shall be used as a LIST_ITEM_HANDLE containing the pending operation. ]*/
+        LIST_ITEM_HANDLE pending_operation_list_item_handle = (LIST_ITEM_HANDLE)context;
+
+        /* Codes_SRS_AMQP_MANAGEMENT_01_169: [ - `on_message_send_complete` shall obtain the pending operation by calling `singlylinkedlist_item_get_value`. ]*/
+        OPERATION_MESSAGE_INSTANCE* pending_operation_message = (OPERATION_MESSAGE_INSTANCE*)singlylinkedlist_item_get_value(pending_operation_list_item_handle);
```
Comment (ericwol-msft) [inline src/amqp_management.c:362]: are you sure there will always be one item on this list?

File: tests/amqp_management_ut/amqp_management_ut.c [hunk 2]
```diff
@@ -1982,6 +1979,7 @@ TEST_FUNCTION(when_on_send_message_complete_indicates_success_it_returns)
     setup_calls_for_pending_operation_with_correlation_id(0);
     (void)amqp_management_execute_operation_async(amqp_management, "some_operation", "some_type", "en-US", test_message, test_on_amqp_management_execute_operation_complete, (void*)0x4244);
```
Comment (ericwol-msft) [inline tests/amqp_management_ut/amqp_management_ut.c:1980]: CLEANUP: replace 0x4244 with test_source_amqp_value ?

File: tests/amqp_management_ut/amqp_management_ut.c [hunk 3]
```diff
@@ -4341,6 +4346,74 @@ TEST_FUNCTION(when_copying_the_string_fails_amqp_management_set_override_status_
...
+    uint64_t correlation_id = 0;
+
+    amqp_management = amqp_management_create(test_session_handle, "test_node");
+    (void)amqp_management_set_override_status_description_key_name(amqp_management, "xxx");
+    (void)amqp_management_open_async(amqp_management, test_on_amqp_management_open_complete, (void*)0x4242, test_on_amqp_management_error, (void*)0x4243);
+    saved_on_message_sender_state_changed(saved_on_message_sender_state_changed_context, MESSAGE_SENDER_STATE_OPEN, MESSAGE_SENDER_STATE_OPENING);
+    saved_on_message_receiver_state_changed(saved_on_message_receiver_state_changed_context, MESSAGE_RECEIVER_STATE_OPEN, MESSAGE_RECEIVER_STATE_OPENING);
+    umock_c_reset_all_calls();
+    setup_calls_for_pending_operation_with_correlation_id(0);
+    (void)amqp_management_execute_operation_async(amqp_management, "some_operation", "some_type", "en-US", test_message, test_on_amqp_management_execute_operation_complete, (void*)0x4244);
```
Comment (ericwol-msft) [inline tests/amqp_management_ut/amqp_management_ut.c:4365]:
```suggestion
    (void)amqp_management_execute_operation_async(amqp_management, "some_operation", "some_type", "en-US", test_message, test_on_amqp_management_execute_operation_complete, test_source_amqp_value );
```

File: tests/amqp_management_ut/amqp_management_ut.c [hunk 4]
```diff
@@ -4341,6 +4346,74 @@ TEST_FUNCTION(when_copying_the_string_fails_amqp_management_set_override_status_
...
+    // arrange
+    AMQP_MANAGEMENT_HANDLE amqp_management;
+    AMQP_VALUE result;
+    int32_t status_code = 200;
+    const char* test_status_description = "my error ...";
+    uint64_t correlation_id = 0;
+
+    amqp_management = amqp_management_create(test_session_handle, "test_node");
+    (void)amqp_management_set_override_status_description_key_name(amqp_management, "xxx");
+    (void)amqp_management_open_async(amqp_management, test_on_amqp_management_open_complete, (void*)0x4242, test_on_amqp_management_error, (void*)0x4243);
```
Comment (ericwol-msft) [inline tests/amqp_management_ut/amqp_management_ut.c:4360]:
```suggestion
    (void)amqp_management_open_async(amqp_management, test_on_amqp_management_open_complete, (void*)0x4242, test_on_amqp_management_error, test_source_amqp_value );
```

File: tests/amqp_management_ut/amqp_management_ut.c [hunk 5]
```diff
@@ -4341,6 +4346,74 @@ TEST_FUNCTION(when_copying_the_string_fails_amqp_management_set_override_status_
...
+    STRICT_EXPECTED_CALL(amqpvalue_create_string("xxx"))
+        .SetReturn(test_status_description_key);
+    STRICT_EXPECTED_CALL(amqpvalue_get_map_value(test_application_properties_map, test_status_description_key))
+        .SetReturn(test_status_description_value);
+    STRICT_EXPECTED_CALL(amqpvalue_get_string(test_status_description_value, IGNORED_PTR_ARG))
+        .CopyOutArgumentBuffer_string_value(&test_status_description, sizeof(test_status_description));
+    STRICT_EXPECTED_CALL(singlylinkedlist_get_head_item(test_singlylinkedlist_handle));
+    STRICT_EXPECTED_CALL(singlylinkedlist_item_get_value(IGNORED_PTR_ARG));
+    STRICT_EXPECTED_CALL(async_operation_cancel(IGNORED_PTR_ARG));
+    STRICT_EXPECTED_CALL(test_on_amqp_management_execute_operation_complete((void*)0x4244, AMQP_MANAGEMENT_EXECUTE_OPERATION_FAILED_BAD_STATUS, 200, "my error ...", test_message));
```
Comment (ericwol-msft) [inline tests/amqp_management_ut/amqp_management_ut.c:4394]:
```suggestion
    STRICT_EXPECTED_CALL(test_on_amqp_management_execute_operation_complete(test_source_amqp_value , AMQP_MANAGEMENT_EXECUTE_OPERATION_FAILED_BAD_STATUS, 200, "my error ...", test_message));
```

**Review by ericwol-msft — APPROVED** [review]

**Comment by ericwol-msft** [comment]
NOTE, there are seg faults on the E2E tests

https://azure-iot-sdks.visualstudio.com/azure-iot-sdks/_build/results?buildId=80842&view=logs&jobId=ec022e2c-bd5d-5a8d-66fe-055600b3609f&j=ec022e2c-bd5d-5a8d-66fe-055600b3609f&t=90a69295-6519-532e-59ab-44b73818dd05

File: src/amqp_management.c (hunk 1 above)
Comment (ewertons) [reply src/amqp_management.c:362]: It should. It is the entry that was created when messagesender_send_async was called. If it is not there, a major memory issue has happened outside the SDK code.

File: src/amqp_management.c [hunk 6]
```diff
@@ -27,7 +27,9 @@ typedef struct OPERATION_MESSAGE_INSTANCE_TAG
     ON_AMQP_MANAGEMENT_EXECUTE_OPERATION_COMPLETE on_execute_operation_complete;
     void* callback_context;
     uint64_t message_id;
+    bool message_sent;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply src/amqp_management.c:30]: That's a good idea. Done.

File: tests/amqp_management_ut/amqp_management_ut.c (hunk 2 above)
Comment (ewertons) [reply tests/amqp_management_ut/amqp_management_ut.c:1980]: So the address of test_source_amqp_value is the same that is being used in this function, but they are different things. The argument in this function is just a void* address that is used as context by the caller. So we should not tie the test_source_amqp_value with that argument, despite the memory addresses chosen being the same.

File: tests/amqp_management_ut/amqp_management_ut.c (hunk 3 above)
Comment (ewertons) [reply tests/amqp_management_ut/amqp_management_ut.c:4365]: So the address of test_source_amqp_value is the same that is being used in this function, but they are different things. The argument in this function is just a void* address that is used as context by the caller. So we should not tie the test_source_amqp_value with that argument, despite the memory addresses chosen being the same.

File: tests/amqp_management_ut/amqp_management_ut.c (hunk 4 above)
Comment (ewertons) [reply tests/amqp_management_ut/amqp_management_ut.c:4360]: So the address of test_source_amqp_value is the same that is being used in this function, but they are different things. The argument in this function is just a void* address that is used as context by the caller. So we should not tie the test_source_amqp_value with that argument, despite the memory addresses chosen being the same.

File: tests/amqp_management_ut/amqp_management_ut.c (hunk 5 above)
Comment (ewertons) [reply tests/amqp_management_ut/amqp_management_ut.c:4394]: So the address of test_source_amqp_value is the same that is being used in this function, but they are different things. The argument in this function is just a void* address that is used as context by the caller. So we should not tie the test_source_amqp_value with that argument, despite the memory addresses chosen being the same.

File: src/amqp_management.c (hunk 1 above)
Comment (ericwol-msft) [reply src/amqp_management.c:362]: No, I mean more then 1.

**Pull request merged**